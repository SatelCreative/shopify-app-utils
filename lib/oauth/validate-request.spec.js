const jsonwebtoken = require('jsonwebtoken');
const mockNow = require('jest-mock-now');
const validateRequest = require('./validate-request');

// Which is 1533160800 UTC
beforeEach(() =>
  mockNow(new Date('Wednesday, 1 August 2018 15:00:00 GMT-07:00')));
afterEach(() => Date.now.mockRestore());

const baseArgs = {
  url: '/',
  jwt: '',
  host: 'https://example.com',
  redirectRoute: '/redirect',
  scopes: ['read_products'],
  key: 'app-key',
  secret: 'hunter2',
  online: false,
  offline: false,
  generateNonce: ({ shop }) => `nonce-${shop}`,
};

describe('oauth/validateRequest', () => {
  it('should redirect as expected in online mode', async () => {
    const onAuthenticated = jest.fn();
    const onRedirect = jest.fn();

    await validateRequest({
      ...baseArgs,
      url: '/?shop=example.myshopify.com',
      jwt: undefined,
      online: true,
      onAuthenticated,
      onRedirect,
    });

    expect(onAuthenticated).toHaveBeenCalledTimes(0);
    expect(onRedirect).toHaveBeenCalledTimes(1);
    expect(onRedirect.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should authenticate as expected in online mode', async () => {
    const onAuthenticated = jest.fn();
    const onRedirect = jest.fn();

    const secret = 'hunter2';
    const jwt = jsonwebtoken.sign(
      {
        shop: 'example.myshopify.com',
        scopes: ['read_products'],
      },
      secret,
    );

    await validateRequest({
      ...baseArgs,
      jwt,
      secret,
      online: true,
      onAuthenticated,
      onRedirect,
    });

    expect(onAuthenticated).toHaveBeenCalledTimes(1);
    expect(onRedirect).toHaveBeenCalledTimes(0);
    // Remove the timestamp
    const args = onAuthenticated.mock.calls[0][0];
    delete args.iat;
    expect(args).toMatchSnapshot();
  });

  it('should not authenticate invalid jwt', async () => {
    const onAuthenticated = jest.fn();
    const onRedirect = jest.fn();

    const secret = 'hunter2';
    const jwt = jsonwebtoken.sign(
      {
        shop: 'example.com',
        scopes: ['read_products'],
      },
      'wrong-secret',
    );

    await validateRequest({
      ...baseArgs,
      jwt,
      secret,
      online: true,
      onAuthenticated,
      onRedirect,
    });

    expect(onAuthenticated).toHaveBeenCalledTimes(0);
    expect(onRedirect).toHaveBeenCalledTimes(0);
  });
});
