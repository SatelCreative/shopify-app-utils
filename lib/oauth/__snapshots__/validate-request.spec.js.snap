// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oauth/validateRequest should authenticate as expected in online mode 1`] = `
Object {
  "decoded": Object {
    "iat": 1533160800,
    "scopes": Array [
      "read_products",
    ],
    "shop": "example.myshopify.com",
  },
  "shop": "example.myshopify.com",
}
`;

exports[`oauth/validateRequest should redirect as expected in online mode 1`] = `
Object {
  "html": "<!DOCTYPE html><html><head><script type=\\"text/javascript\\">window.top.location.href = \\"https://example.myshopify.com/admin/oauth/authorize?client_id=app-key&state=nonce-example.myshopify.com&redirect_uri=https%3A%2F%2Fexample.com%2Fredirect&scope=read_products&grant_options%5B%5D=per-user\\"</script></head></html>",
  "url": "https://example.myshopify.com/admin/oauth/authorize?client_id=app-key&state=nonce-example.myshopify.com&redirect_uri=https%3A%2F%2Fexample.com%2Fredirect&scope=read_products&grant_options%5B%5D=per-user",
}
`;
